### ISSUE TITLE
Confusing error message when using apt_key with a url 

### ISSUE THREAD

**Issue by ashak** [ansible/ansible-modules-core#3246]
##### Issue Type:
- Bug Report
##### Plugin Name:

apt_key module specifically when using url
##### Ansible Version:

```
ansible 2.1.0 (devel cc0d58abcc) last updated 2016/03/11 20:18:43 (GMT +100)
  lib/ansible/modules/core: (detached HEAD c86a0ef84a) last updated 2016/03/11 20:21:07 (GMT +100)
  lib/ansible/modules/extras: (detached HEAD 33a557cc59) last updated 2016/03/11 20:18:52 (GMT +100)
  config file = /home/garyr/code/personal-ansible/ansible.cfg
  configured module search path = Default w/o overrides
```
##### Ansible Configuration:

inventory=${ANSIBLE_HOME}/contrib/inventory/ec2.py
##### Environment:

Running on Debian Stretch, managing Debian Jessie.
##### Summary:

I used apt_key to try to install rspamd's gpg key based on the URL they provide on their site. WHen Ansible tried to download the key it gave me this error 'Failed to download key at https://rspamd.com/apt-stable/gpg.key: Connection failure: [X509] PEM lib'. I tested downloading of the file using curl on machine being managed and it downloaded fine without issue. There was no 'connection failure' as ansible suggested. I eventually tried validate_cert: no.... and Ansible was successfully able to download the key. My guess on the thing causing the error is SNI related (based on reading older issues and seeing that without SNI the server at rspamd.net serves up pastie.rspamd.net by default). But I am more concerned that the error I received suggested the problem was a connection failure, not a mismatched/invalid certificate.
##### Steps To Reproduce:

```
- name: add the rspamd apt key
  apt_key:
    url: https://rspamd.com/apt-stable/gpg.key
```
##### Expected Results:

The error I receive should indicate the problem is with the certificate and not with a connection failure
##### Actual Results:

```
TASK [add the rspamd apt key] ************************************
fatal: [a.b.c.d]: FAILED! => {"changed": false, "failed": true, "msg": "Failed to download key at https://rspamd.com/apt-stable/gpg.key: Connection failure: [X509] PEM lib (_ssl.c:2751)"}

```

**Comment by alikins** [comment]
Looks like SNI support is dependent on python 2.7.9 on the managed client. What version of python is on the managed client? (looks like Jessie should be 2.7.9+).

Would it be possible to get a '-vvvvv' debug log output from this failure?

**Comment by nsteinmetz** [comment]
Hi,

I have the same on jessie server with:

Action:

```
- name: Add InfluxData Debian Repo key
  apt_key: url=https://repos.influxdata.com/influxdb.key  state=present
```

Result:

```
fatal: [xxx.xxx.xxx.xxx]: FAILED! => {"changed": false, "failed": true, "msg": "Failed to download key at https://repos.influxdata.com/influxdb.key: Connection failure: unknown error (_ssl.c:2751)"}
```

Python on Jessie is 2.7.9

How can I provde the -vvvvvv debug log output ?

I tried with:

```
- name: Add InfluxData Debian Repo key
  apt_key: url=https://repos.influxdata.com/influxdb.key  state=present
  register: result

- debug: var=result
```

But it does not work.

Thanks,
Nicolas

**Comment by nsteinmetz** [comment]
And I have the same result with:

```
- name: Download telegraf
  get_url: url=https://repos.influxdata.com/influxdb.key  dest=/tmp/influxdb.key mode=0440
  tags: telegraf
```

Output:

```
fatal: [xxx.xxx.xxx.xxx]: FAILED! => {"changed": false, "dest": "/tmp/influxdb.key", "failed": true, "msg": "Request failed", "response": "Connection failure: unknown error (_ssl.c:2751)", "state": "absent", "status_code": -1, "url": "https://repos.influxdata.com/influxdb.key"}
```

It's not strictly linked with `apt_key` action.

**Comment by nsteinmetz** [comment]
Hmm seems if's fixed once `wget` or more probably `ca-certificates` are installed.

**Comment by tumbl3w33d** [comment]
I can confirm that installing ca-certificates on the target host solved the issue for me.

**Comment by Hashfyre** [comment]
I'm having this same issue with the following task:

```
  - name: Add Duo repo public signing key
    apt_key:
      url: https://duo.com/APT-GPG-KEY-DUO
      id: 1CC91FC615D32EFC
      state: present
```

Installing `ca-certificates` did not solve the issue for me.

**Comment by ansibot** [comment]
@jvantuyl, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by jvantuyl** [comment]
This is really an upstream Python issue.  Not much I can do except recommend that you try what's already been mentioned.  @Hashfyre What version of Python are you using?

In general, though, this is a wontfix (because I can't).

**Comment by azaghal** [comment]
Apologies for posting on closed/dead issue, but it might save someone a lot of headache to know another possible cause for the problem described here.

In my particular case, the error was happening due to having a malformed certificate file in `/etc/ssl/certs/` directory (on Debian Jessie, with Ansible 1.9.6). In particular, I had two certificates deployed in one file, and a line `------END CERTIFICATE----------BEGIN CERTIFICATE-----` (i.e. the missing newline between first and second certificate). Once the file has been fixed, things started working again.
